jerry-main/fuzzilli.c

```diff
@@ -0,0 +1,376 @@
+
+struct shmem_data
+{
+  uint32_t num_edges;
+  unsigned char edges[];
+};
+
+struct shmem_data *__shmem;
+uint32_t *__edges_start, *__edges_stop;
+
+void __sanitizer_cov_reset_edgeguards()
+{
+  uint32_t N = 0;
+  for (uint32_t *x = __edges_start; x < __edges_stop && N < MAX_EDGES; x++)
+    *x = ++N;
+}
+
+void __sanitizer_cov_trace_pc_guard_init(uint32_t *start, uint32_t *stop)
+{
+  // Avoid duplicate initialization
+  if (start == stop || *start)
+    return;
+
+  if (__edges_start != NULL || __edges_stop != NULL)
+  {
+    fprintf(stderr, "Coverage instrumentation is only supported for a single module\n");
+    _exit(-1);
+  }
+
+  __edges_start = start;
+  __edges_stop = stop;
+
+  // Map the shared memory region
+  const char *shm_key = getenv("SHM_ID");
+  if (!shm_key)
+  {
+    puts("[COV] no shared memory bitmap available, skipping");
+    __shmem = (struct shmem_data *)malloc(SHM_SIZE);
+  }
+  else
+  {
+    int fd = shm_open(shm_key, O_RDWR, S_IRUSR | S_IWUSR);
+    if (fd <= -1)
+    {
+      fprintf(stderr, "Failed to open shared memory region: %s\n", strerror(errno));
+      _exit(-1);
+    }
+
+    __shmem = (struct shmem_data *)mmap(0, SHM_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+    if (__shmem == MAP_FAILED)
+    {
+      fprintf(stderr, "Failed to mmap shared memory region\n");
+      _exit(-1);
+    }
+  }
+
+  __sanitizer_cov_reset_edgeguards();
+
+  __shmem->num_edges = (uint32_t)(stop - start);
+  printf("[COV] edge counters initialized. Shared memory: %s with %u edges\n", shm_key, __shmem->num_edges);
+}
+
+void __sanitizer_cov_trace_pc_guard(uint32_t *guard)
+{
+  // There's a small race condition here: if this function executes in two threads for the same
+  // edge at the same time, the first thread might disable the edge (by setting the guard to zero)
+  // before the second thread fetches the guard value (and thus the index). However, our
+  // instrumentation ignores the first edge (see libcoverage.c) and so the race is unproblematic.
+  uint32_t index = *guard;
+  __shmem->edges[index / 8] |= 1 << (index % 8);
+  *guard = 0;
+}
+
+// We have to assume that the fuzzer will be able to call this function e.g. by
+// enumerating the properties of the global object and eval'ing them. As such
+// this function is implemented in a way that requires passing some magic value
+// as first argument (with the idea being that the fuzzer won't be able to
+// generate this value) which then also acts as a selector for the operation
+// to perform.
+jerry_value_t
+jerryx_handler_fuzzilli(const jerry_value_t func_obj_val, /**< function object */
+                        const jerry_value_t this_p,       /**< this arg */
+                        const jerry_value_t args_p[],     /**< function arguments */
+                        const jerry_length_t args_cnt)    /**< number of function arguments */
+{
+  (void)func_obj_val; /* unused */
+  (void)this_p;       /* unused */
+
+  jerry_char_t operation[256] = {0};
```

Comment:
Is there any reason of limitation? IMO, you can use `jerry_heap_alloc` at this point then no size check would be needed below.